range_queries/fenwick_tree_2d.cpp

```diff
@@ -0,0 +1,129 @@
+#include <iostream>
+#include <cassert>
+#include <vector>
+
+
+template <class T>
+struct fenwick {
+    // siz - size of the fenwick tree
+    // arr - fenwick tree array
+    int siz;
+    std::vector<T> arr;
+
+    // Constructor 
+    // n - size of the array on which fenwick tree is used
+    // creates a new array with all entries as 0
+    // indicies are 1-indexed
+    explicit fenwick(int n) {
+	siz = n;
+	arr.resize(n + 1);
+    }
+
+    // Calculates a prefix sum till index idx
+    T sum(int idx) {
+	T an = 0;
+	while(idx > 0) {
+	    an += arr[idx];
+	    // subtract first active bit from idx
+	    idx -= idx & -idx;
+	}
+	return an;
+    }
+
+    // update the values at idx by delta
+    // update all the indices which include idx in their sum
+    void update(int idx, T delta) {
+	while (idx < siz + 1) {
+	    arr[idx] += delta;
+	    // add the first active bit to idx
+	    idx += idx & -idx;
+	}
+    }
+
+    // find sum in range l to r
+    T query(int l, int r) {
```

Comment:
finally, here are even more implicit assumptions on the input, `0 < l <= r <= siz` (I believe)